libp2p/transports/transport.nim

```diff
@@ -21,27 +21,66 @@ logScope:
   topics = "libp2p transport"
 
 type
+  ListenErrorCallback* = proc (
+      ma: MultiAddress,
+      err: ref CatchableError): Future[ref TransportListenError]
+      {.gcsafe, raises: [Defect].}
   TransportError* = object of LPError
   TransportInvalidAddrError* = object of TransportError
   TransportClosedError* = object of TransportError
+  TransportListenError* = object of TransportError
+    ma*: MultiAddress
 
   Transport* = ref object of RootObj
     addrs*: seq[Multiaddress]
     running*: bool
     upgrader*: Upgrade
+    listenError*: ListenErrorCallback
 
 proc newTransportClosedError*(parent: ref Exception = nil): ref LPError =
   newException(TransportClosedError,
     "Transport closed, no more connections!", parent)
 
+proc newTransportListenError*(
+    ma: MultiAddress,
+    parent: ref Exception = nil): ref TransportListenError =
+
+  return (ref TransportListenError)(msg: "Transport failed to start", parent: parent, ma: ma)
+
+const ListenErrorDefault* =
+  proc(
+      ma: MultiAddress,
+      err: ref CatchableError): Future[ref TransportListenError] {.async.} =
+
+    return newTransportListenError(ma, err)
+
+method handles*(
+  self: Transport,
+  address: MultiAddress): bool {.base, gcsafe.} =
+  ## check if transport supports the multiaddress
+  ##
+
+  # by default we skip circuit addresses to avoid
+  # having to repeat the check in every transport
+  if address.protocols.isOk:
+    return address.protocols
+    .get()
+    .filterIt(
+      it == multiCodec("p2p-circuit")
+    ).len == 0
+
 method start*(
   self: Transport,
   addrs: seq[MultiAddress]) {.base, async.} =
   ## start the transport
   ##
 
   trace "starting transport on addrs", address = $addrs
-  self.addrs = addrs
+  self.addrs = addrs.filter(proc(ma: MultiAddress): bool =
+    let handled = self.handles(ma)
+    if not handled:
+      trace "Invalid address detected, skipping!", address = ma
```

Comment:
This should never happen, since we filter in switch.start, no?

Reply:
Hmm good point. `tcptransport.start` had logic to skip unhandled addresses, but I guess that can just be removed instead of moving it to the transport base class.